smslibrary/src/test/java/com/eis/smslibrary/SMSDeliveredBroadcastReceiverTest.java

```diff
@@ -0,0 +1,65 @@
+package com.eis.smslibrary;
+
+import android.app.Activity;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+
+import com.eis.smslibrary.listeners.SMSDeliveredListener;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+public class SMSDeliveredBroadcastReceiverTest {
+    private SMSDeliveredBroadcastReceiver receiver;
+    private SMSDeliveredListener listenerMock = mock(SMSDeliveredListener.class);
+    private Context contextMock = mock(Context.class);
+    private final SMSPeer VALID_PEER = new SMSPeer("+393423541601");
+    private final String VALID_TEXT = "Some random text to send";
+    private final SMSMessage VALID_MESSAGE = new SMSMessage(VALID_PEER, VALID_TEXT);
+
+    @Before
+    public void setup(){
+        receiver = spy(new SMSDeliveredBroadcastReceiver(VALID_MESSAGE, listenerMock));
+    }
+
+    @Test
+    public void constructor_doesntThrow(){
+        new SMSDeliveredBroadcastReceiver(VALID_MESSAGE, listenerMock);
+    }
+
+    @Test
+    public void receive_success_correctListener(){
+        doReturn(Activity.RESULT_OK).when((BroadcastReceiver)receiver).getResultCode();
+        receiver.onReceive(contextMock, null);
+        verify(listenerMock, times(1))
+                .onSMSDelivered(VALID_MESSAGE, SMSMessage.DeliveredState.MESSAGE_DELIVERED);
+        verify(contextMock).unregisterReceiver(receiver);
```

Comment:
Well, now I try to write what you did to understand well, tell me if I'm wrong:
1) Force the `getResultCode ()` method, once called, to return `Activity.RESULT_OK`
2) Invoke the `onReceive` method which should assign the` MESSAGE_DELIVERED` value to the `DeliveryState` variable
3) Check that the `onSMSDelivered` method has been called
4) Check that the `unregisterReceiver` method has been called

I don't quite understand what the `times (x)` parameter is for...

Reply:
yup, you're definitely right. When `getResultCode()` is called it will automatically return `Activity.RESULT_OK`
(**N.B.**, I want the `BroadcastReceiver` to do so, not the `SMSDelibveredBroadcastReceiver`, so I need that cast.
then I'll call the function and verify is eveything went according to plans

The `times(1)` is used to check how many times the function was called, so `times(5)` will check if the function was called 5 times

Reply:
Another insteresting thing. `SMSSentBroadcastReceiver` (and Delivered too) are **NOT** mocks, they are Spies, the difference is that a mock is an object which when a non-mocked function is called nothing is returned or done, while a spy is an object which will work as a normal oobject but where I can **replace** functions with mocks